### PR TITLE
READY: join impl

### PR DESCRIPTION
**Pull request by Sakapoi**

### CODE REVIEW

File: file/join.rs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+    #[arg(short, long, value_delimiter = ',')]
+    o: Option<Vec<String>>,
+
+    /// Field separator character
+    #[arg(short, long, default_value_t = ' ')]
+    t: char,
+
+    /// Output only unpairable lines from file_number
+    #[arg(short, long)]
+    v: Option<u8>,
```
Comment (0x07C0) [inline file/join.rs:44]: Fields for argument names should be descriptive. No need for `long` args here,

File: file/join.rs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    /// Join on the specified field of file 1
+    #[arg(short = '1', long, default_value_t = 1)]
+    field1: usize,
+
+    /// Join on the specified field of file 2
+    #[arg(short = '2', long, default_value_t = 1)]
+    field2: usize,
```
Comment (0x07C0) [inline file/join.rs:52]: This args don't have long variants in linux util.

File: file/join.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+        .map(|s| s.to_string())
+        .collect()
+}
+
+fn read_file_lines(file_path: &PathBuf, sep: char) -> Result<Vec<Vec<String>>, Box<dyn std::error::Error>> {
+    let file = File::open(file_path).map_err(|_| format!("Unable to open file"))?;
+    let reader = BufReader::new(file);
+    let mut lines = Vec::new();
+
+    for line in reader.lines() {
```
Comment (0x07C0) [inline file/join.rs:72]: Reading the whole file will result in out of memory and crush the app. Reading from files needs to be heavily refactored

File: file/join.rs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        }
+    }
+
+    if a.unwrap_or(0) == 2 {
+        for (key, line1) in &map2 {
+            if !map1.contains_key(key) {
+                println!("{}", line1.join(" "));
+            }
+        }
+    }
```
Comment (0x07C0) [inline file/join.rs:195]: Just set default values for those args, so you don't have to unwrap with default every time.

File: file/join.rs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+    file2: Vec<Vec<String>>,
+    field1: usize,
+    field2: usize,
+    a: Option<u8>,
+    e: Option<String>,
+    o: Option<Vec<String>>,
+    v: Option<u8>,
+) -> Result<(), Box<dyn std::error::Error>>{
+    if field1 == 0 || field2 == 0 {
+        return Err(format!("field 1 and field 2 must be greater than one").into());
```
Comment (0x07C0) [inline file/join.rs:91]: Why do we need `format!` and `.into()` here? Also, f1 and f2 can be equal to 1 too, not greater then.

File: file/join.rs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    if o.is_some() {
+        let mut i = 0;
+        for s in &file1 {
+            let mut res = Vec::new();
+            for num in o.clone().unwrap() {
```
Comment (0x07C0) [inline file/join.rs:98]: You shouldn't use is_some followed by unwrap. Use `if let Some()`

File: file/join.rs [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+fn parse_fields(line: &str, sep: char) -> Vec<String> {
+    line.split(sep)
+        .map(|s| s.to_string())
+        .collect()
+}
+
+fn read_file_lines(file_path: &PathBuf, sep: char) -> Result<Vec<Vec<String>>, Box<dyn std::error::Error>> {
+    let file = File::open(file_path).map_err(|_| format!("Unable to open file"))?;
```
Comment (0x07C0) [inline file/join.rs:68]: You have not accounted for stdin as an input.

File: file/join.rs [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+        for s in &file1 {
+            let mut res = Vec::new();
+            for num in o.clone().unwrap() {
+                let f_num: Vec<&str> = num.split('.').collect();
+                if f_num[0] == "1" {
+                    res.push(s[f_num[1].parse::<usize>().unwrap() - 1].clone());
+                }
+                if f_num[0] == "2" {
+                    if i >= file2.len() && e.is_some() {
+                        res.push(e.clone().unwrap());
```
Comment (0x07C0) [inline file/join.rs:105]: is_some followed by unwrap

File: file/join.rs [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    if o.is_some() {
+        let mut i = 0;
+        for s in &file1 {
+            let mut res = Vec::new();
+            for num in o.clone().unwrap() {
+                let f_num: Vec<&str> = num.split('.').collect();
+                if f_num[0] == "1" {
+                    res.push(s[f_num[1].parse::<usize>().unwrap() - 1].clone());
```
Comment (0x07C0) [inline file/join.rs:101]: Don't unwrap on usize parse

File: file/join.rs [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+    a: Option<u8>,
+    e: Option<String>,
+    o: Option<Vec<String>>,
+    v: Option<u8>,
+) -> Result<(), Box<dyn std::error::Error>>{
+    if field1 == 0 || field2 == 0 {
+        return Err(format!("field 1 and field 2 must be greater than one").into());
+    }
+
+    if o.is_some() {
```
Comment (0x07C0) [inline file/join.rs:94]: This whole block looks like it could've been a separate function.

File: file/join.rs [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+
+                output.extend_from_slice(&l);
+                output.extend_from_slice(&line[1..]);
+
+                if v.unwrap_or(0) != 1 && v.unwrap_or(0) != 2 {
+                    println!("{}", output.join(" "));
+                }
+            }
+        } else if a.unwrap_or(0) == 2 {
+            let mut output = vec![String::from("(unknown)")];
```
Comment (0x07C0) [inline file/join.rs:146]: Empty fields by default stay empty. `-e` is used to replace them with a custom filed.

File: file/tests/join/mod.rs [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+fn field2_test() {
+    let project_root = env!("CARGO_MANIFEST_DIR");
+    let file1 = format!("{}/tests/join/file1.txt", project_root);
+    let file2 = format!("{}/tests/join/file2.txt", project_root);
+    let args = ["-2", "1", file1.as_str(), file2.as_str()];
+
+    let expected_output = "1 Alice HR\n2 Bob Finance\n3 Charlie IT\n";
+
+    run_test_join(&args, &expected_output, "", 0)
+}
```
Comment (0x07C0) [inline file/tests/join/mod.rs:124]: Both tests represent default behavior.

File: file/join.rs [hunk 13]
```diff
@@ -0,0 +1,224 @@
...
+
+    for line in &file2 {
+        let key = line[field2 - 1].clone();
+        if let Some(matches) = map.get_mut(&key) {
+            matched.insert(key.clone(), true);
+            for l in matches.iter_mut() {
+                let mut output = vec![];
+
+                output.extend_from_slice(&l);
+                output.extend_from_slice(&line[1..]);
```
Comment (0x07C0) [inline file/join.rs:139]: Will this work for cases where the key is not the first column in the file? Also wouldn't doing it in a loop duplicate the data for each match?

**Review by 0x07C0 — CHANGES_REQUESTED** [review]
Please, rework your data structures too, `H<S, V<V<S>>>` can't be optimal. And find a way to minimize cloning, Almost every string gets cloned multiple times.

File: file/join.rs [hunk 14]
```diff
@@ -404,13 +417,119 @@ fn process_files(
...
+    } else {
+        Box::new(BufReader::new(File::open(file2_path)?))
+    };
+
+    let mut file2_cache = vec![];
+    for line in file2.lines() {
+        let line = line?;
+        let fields2 = parse_fields(&line, sep);
+        file2_cache.push(fields2);
+    }
```
Comment (SupperZum) [inline file/join.rs:451]: file2_cache can lead to memory overflow and crash the app. Correct me if I'm wrong.

**Review by SupperZum — CHANGES_REQUESTED** [review]

**Review by SupperZum — APPROVED** [review]

**Pull request merged**